libmamba/include/mamba/core/solution.hpp

```diff
@@ -0,0 +1,202 @@
+// Copyright (c) 2023, QuantStack and Mamba Contributors
+//
+// Distributed under the terms of the BSD 3-Clause License.
+//
+// The full license is in the file LICENSE, distributed with this software.
+
+#ifndef MAMBA_CORE_SOLUTION_HPP
+#define MAMBA_CORE_SOLUTION_HPP
+
+#include <variant>
+#include <vector>
+
+#include "package_info.hpp"
+
+namespace mamba
+{
+    class Solution
+    {
+    public:
+
+        struct Upgrade
+        {
+            PackageInfo remove;
+            PackageInfo install;
+        };
+        struct Downgrade
+        {
+            PackageInfo remove;
+            PackageInfo install;
+        };
+        struct Change
+        {
+            PackageInfo remove;
+            PackageInfo install;
+        };
+        struct Reinstall
+        {
+            PackageInfo what;
+        };
+        struct Remove
+        {
+            PackageInfo remove;
+        };
+        struct Install
+        {
+            PackageInfo install;
+        };
+        using Action = std::variant<Upgrade, Downgrade, Change, Reinstall, Remove, Install>;
+        using action_list = std::vector<Action>;
+
+        using size_type = typename action_list::size_type;
+
+        Solution() = default;
+        Solution(const action_list& actions);
+        Solution(action_list&& actions);
+
+        [[nodiscard]] auto empty() const -> bool;
+        [[nodiscard]] auto size() const -> size_type;
+
+        [[nodiscard]] auto actions() const -> const action_list&;
+        [[nodiscard]] auto actions() -> action_list&;
```

Comment:
Why allow modifying the action list?

Reply:
In general I would say, there is no invariant the this violates, so why not make it writable too.

In particular, there is currently some logic that needs to appends more entries to the actions, such as adding Python noarch packages for reinstall.
Hopefully I can change mamba behavior for this to be computed before creating this class, but this would be outside this PR.